### DISCUSSION TITLE
Working together with `@capacitor-community/barcode-scanner`

### DISCUSSION THREAD

**Discussion by tafelnl** [capawesome-team/capacitor-mlkit#45] [General]
As the creator and previous maintainer of `@capacitor-community/barcode-scanner` I would like to mediate between you and @thegnuu to talk about the future of both packages. Although competition is good on the market, I think with software plugins this is a little different. I think it's in the best interest of the community to join forces and maintain a shared plugin. Especially since @thegnuu and others have put a large effort in implementing the ML Kit library as well. I guess the main question would be under which name the project will continue. @robingenz as you didn't submit a PR or anything to `@capacitor-community/barcode-scanner`, I guess you might already have ideas about your preferred course of action?

**Comment by thegnuu** [comment]
@tafelnl Thank you for pointing out, that this repo exists.

@robingenz I think it would be a good idea to join our forces and work on one plugin for the community since the problem that has to be solved is literally the same and the upcoming maintenance work to update the dependencies aso. will be the same as well. Just by looking at the download stats of npm, it might make sense to work on `@capacitor-community/barcode-scanner` since the user base seems quite higher with 18k weekly downloads.

I realized you have a few features the community was asking for on my side as well, like defining a detection area and scanning codes on an existing image.

Just let me know what you think!

**Reply by robingenz** [reply]
@tafelnl Thank you for bringing this up and for all your work in the Capacitor community.

@thegnuu Thanks to you as well for your work. I would love to work together on a plugin because as you mentioned, we need to do similar maintenance work in the future and we should not double the work. I started with the development of the barcode scanning plugin already last year. First as a single plugin in a private repository for an internal project, then as a mono repo in a public repository. The `@capacitor-community/barcode-scanner` plugin was unfortunately not an option to start with, because I needed an implementation with the ML Kit SDK from the beginning. It was important to me to develop the plugin from scratch without annoying users with breaking changes. The two projects differ in various ways for this reason.

Currently I see the following benefits with `@capacitor-mlkit/barcode-scanning`:
1. **Consistent versioning**: The ML Kit has many SDKs, not only barcode scanning. As soon as you use more than one Capacitor ML Kit plugin in your project, you have to keep the ML Kit SDK version in sync with other ML Kit SDK versions of other Capacitor plugins. You have the same problem with the Firebase SDK, which is why I created [capacitor-firebase](https://github.com/capawesome-team/capacitor-firebase). This problem is solved by this repository. Many more Capacitor ML Kit plugins will come this year.
1. **More features**: As you mentioned, the Capawesome plugin has several features that the Capacitor Community plugin currently lacks. These include the definition of the detection area and the scanning of barcodes on images. However, I would like to highlight the optional *ready-to-use interface*, for which no WebView customizations are necessary and whose implementation was very extensive.
1. **Easier start**: Thanks to the optional ready-to-use interface, the plugin can be tried out immediately with a few lines of code without any WebView customizations.
1. **Financial support**: The development of this and all other Capawesome plugins is currently supported by many sponsors on GitHub and Open Collective. It is planned that in the future (later this year) bounties can be placed on issues so that the community has better motivation to contribute to the project. These bounties will then be paid out via the Capawesome profile on Open Collective.
1. **Blog posts and guides**: There are already blog posts and guides that provide an easy introduction to using the Capawesome plugin, see [Announcing the Capacitor ML Kit Barcode Scanning Plugin](https://capawesome.io/blog/announcing-the-capacitor-mlkit-barcode-scanner-plugin/) and [How to build an Ionic Barcode Scanner with Capacitor](https://ionic.io/blog/how-to-build-an-ionic-barcode-scanner-with-capacitor).
1. **Demo apps**: There are already two demo apps that show users how to use the Capawesome plugin and allow direct testing without creating an app yourself.

For these reasons, I think it makes more sense to continue working on `@capacitor-mlkit/barcode-scanning`. It has significantly less npm downloads at the moment, but it is also quiet new.

Feel free to let me know what you think.

**Reply by thegnuu** [reply]
This seems like a totally valid description of why we should stick to this repo, especially the part with the consistent ML Kit versioning which could cause a lot of issues, I have not even thought about that. Of course, the other things you are pointing out are right as well. I would like to introduce @JanMisker to this discussion since he did a lot of work on the ML-Kit implementation in `@capacitor-community/barcode-scanner` and see what he thinks.

A ready-to-use UI would have been one of the next things I had to implement next since issues regarding the proper setup of the webView are 90% of the issues in my inbox.

One thing I see is missing is the web implementation, but I have planned to rework this implementation anyway on my side and I know that some users were using this part since the introduction in `v4.0.0`, so I think this would be required to motivate some users to switch to `@capacitor-mlkit/barcode-scanning`. But since I had planned to do this anyway, I can do this here as well.

In my opinion, the most important thing is that we find a way to work together, I don't really care if this results in deprecating `@capacitor-community/barcode-scanner` and maintaining `@capacitor-mlkit/barcode-scanning` or the other way around :)

**Reply by JanMisker** [reply]
@thegnuu thanks for adding me to this discussion. I agree with you that, even though 'our' plugin has a bigger user-base, the 'new kid on the block' seems better to move forward with. Biggest reason being that it already works with ML-Kit :) 

I have to say I'm using the current beta version in a production app and that works great, so in a sense we're not far off from release. However we would introduce breaking API changes, so it would anyway require a rewrite for the users of it. Plus I like the `addListener()` pattern of this plugin.

**Reply by robingenz** [reply]
@thegnuu @JanMisker Thanks a lot for your feedback.

> A ready-to-use UI would have been one of the next things I had to implement next since issues regarding the proper setup of the webView are 90% of the issues in my inbox.

Yes, I think for most a ready-made UI is all you need and it makes it extremely easy to get started with the plugin.

> One thing I see is missing is the web implementation, but I have planned to rework this implementation anyway on my side and I know that some users were using this part since the introduction in v4.0.0, so I think this would be required to motivate some users to switch to @capacitor-mlkit/barcode-scanning. But since I had planned to do this anyway, I can do this here as well.

Yes, there is already an issue for this: #20
I have also created an implementation with the [Barcode Detection API](https://developer.mozilla.org/en-US/docs/Web/API/Barcode_Detection_API): https://github.com/robingenz/ionic-capacitor-barcode-scanner/blob/728f810c6c22ca5aee45c480573b1fb36118bfc0/src/app/home/web-barcode-scanner.service.ts

> In my opinion, the most important thing is that we find a way to work together, I don't really care if this results in deprecating @capacitor-community/barcode-scanner and maintaining @capacitor-mlkit/barcode-scanning or the other way around :)

Of course! I would love to see you contribute to the project. PRs are always welcome. I always try to review and release quickly. For more complex issues it might be useful to discuss implementation details in advance.

**Reply by thegnuu** [reply]
@robingenz @JanMisker and @tafelnl This sounds like an idea, what do you think would be an "ideal" way of letting users know they should migrate to this plugin? I would like to implement the web platform prior to this, but once this is done I think it would make sense to post a deprecation notice in `@capacitor-community/barcode-scanner` and let them know that we will work together on this plugin in the future.

As @JanMisker mentioned the timing is ideal as well, since the upcoming version would have been breaking for all users anyways.

**Reply by tafelnl** [reply]
Great to hear that you guys are all positive about this :smile: 

NPM and GitHub both have a feature to mark the package as 'deprecated'. Then in the README a more elaborate message can be added.

Only thing I am just a bit hesitant about is that the `@capacitor-community` is under control of the official Ionic Capacitor team and `@capacitor-mlkit` is maintained by Capawesome (@robingenz). Although I am sure you have no weird intentions or anything like that @robingenz ;) it's still a bit weird maybe to transfer the 'ownership' to a 'third party' so to say. I am also absolutely not against it. In the contrary, I think that you (@robingenz) are doing a very nice job in creating and maintaining a bunch of top of the class Capacitor plugins. And earning money off of that seems only fair to me. But I think it would be justified to maybe make some kind of statement, so users of the `@capacitor-community` know what they're up to. Because without reading to much into it, I can imagine it seems like a weird change for a user to switch from a package somewhat owned by the Ionic team to a complete third-party one.

**Reply by JanMisker** [reply]
Indeed I think that Ionic/Capacitor should be involved as well in any transfer/deprecation. From my (our?) perspective it would simply be an abandonment of the `@capacitor/community` plugin. Then at some point Ionic would notice and they could put a deprecation notice. Of course the readme could be updated by us to direct users to a new plugin, but it feels to me that a formal deprecation should be done by the managers of the `@capacitor/community` namespace. Maybe @tafelnl or @thegnuu could reach out to the Ionic managers?

Another possibility could be to go the other way around; `@capacitor-mlkit` would be 'deprecated' as a plugin and transferred to `@capacitor-community`, basically version 5 could be a 'complete rewrite' aka copy of `@capacitor-mlkit`. But I guess that doesn't align well with @robingenz's ideas about how he wants to maintain the package?

**Reply by thegnuu** [reply]
I think it is a good idea to reach out to `@capacitor-community` first to check things out.

I was in contact with @thomasvidas back in February 2022 when I joined as a maintainer, maybe he can help us out here or point to someone who is able to?

**Reply by tafelnl** [reply]
Thomas isn't working at Capacitor anymore AFAIK. But Capacitor used to have Slack (not sure if they'll still respond there as they moved away from it a while back). And now they have Discord: https://ionic.link/discord

**Reply by JanMisker** [reply]
A few weeks ago @jcesarmobile checked in on the repository to see whether is maintained, maybe he can pitch in?

**Reply by robingenz** [reply]
> it's still a bit weird maybe to transfer the 'ownership' to a 'third party' so to say.

I see what you mean but the same was done very successfully with ionic-native.

> Indeed I think that Ionic/Capacitor should be involved as well in any transfer/deprecation.

Yes, I think so too. I would like to hear the opinion of the Capacitor team.

> Another possibility could be to go the other way around; `@capacitor-mlkit` would be 'deprecated' as a plugin and transferred to `@capacitor-community`

Yes, that would not be in my interest. For one thing, this is not a single plugin but a mono repo with multiple plugins with their own npm namespace. For another, I appreciate the possibilities this GitHub organization gives me.

**An alternative would be**: We keep both plugins, but not both plugins use the same implementation. There are also reasons not to use the ML Kit SDK. I've had several requests now expressing concerns about privacy or build size. The Capacitor community barcode scanner plugin has been very popular so far. We may just leave the implementation there as is and give users the option to choose which plugin is right for them.

**Reply by JanMisker** [reply]
@robingenz the reason we want to switch to ML-Kit is because the current implementation is based on an ancient zxing QR decoding library which is poorly maintained and seems to cause issues on newer Android devices.

**Reply by thegnuu** [reply]
@robingenz @JanMisker  I don't think that keeping `@capacitor-community/barcode-scanner` and maintaining it in the current state is a good idea. The current implementation is missing some features on both sides, iOS and Android, that would require a great effort to implement (scan area and stuff like this), and there would still be the issue that the results and performance using `zxing` and `AVFoundation` are not really usable, especially if you want to scan 1D-Codes. This is the main reason we decided to move on with ML-Kit for v5.

I don't think the privacy aspect is a big deal since all the processing happens on the device. I am not sure about the impacts on build size, I have never checked that, but I believe that this will not be an issue for the major part of the user base...

**Comment by thegnuu** [comment]
Thanks @tafelnl, I will reach out to them via Discord!

**Comment by mhartington** [comment]
Hey folks! Taking a look over the comments here, just to make sure I understand the plan, the plan is to...

1) Create a new `@capacitor-mlkit` or with `@capacitor-mlkit/barcode-scanner` as one of the plugins
2) Deprecate `@capacitor-community/barcode-scanner` on github and npm and point to `@capacitor-mlkit/barcode-scanner`
3) Update any docs references on the ionic/capacitor side to point to the new plugin.

Is this correct?

**Reply by robingenz** [reply]
Yes, that is currently the idea. But I have just posted an alternative above (see https://github.com/capawesome-team/capacitor-mlkit/discussions/45#discussioncomment-6153440).

**Reply by thegnuu** [reply]
Thanks for reaching out @mhartington!

The repo and npm package already exists (the one where this discussion is held), it is a mono-repo including the npm packages for the org `@capacitor-mlkit`. The current plan would be to keep this repo as is and then follow steps 2. and 3. you mentioned.

To clarify:

1. Keep this repo as is to hold the new code base
2. Deprecate `@capacitor-community/barcode-scanner` on github and npm and point to `@capacitor-mlkit/barcode-scanner`
3. Update any docs references on the ionic/capacitor side to point to the new plugin.
